### DISCUSSION TITLE
引用特性是否可以完全去除

### DISCUSSION THREAD

**Discussion by zhangzhimin** [prajna-lang/prajna#19] [Programming Language Design]
般若目前把类似C++的引用传递, 引用变量, 返回值引用类型等都取消了.

时候存在必须使用引用更为合适的场景?

**Comment by microctar** [comment]
# *[Wikipedia: Value type and reference type](https://en.wikipedia.org/wiki/Value_type_and_reference_type)*

Reference types are represented as a reference to another value, which may itself be either a value or reference type. Reference types are often implemented using [pointers](https://en.wikipedia.org/wiki/Pointer_(computer_programming)), though many high-level programming languages such as [Java](https://en.wikipedia.org/wiki/Java_(programming_language)) and [Python](https://en.wikipedia.org/wiki/Python_(programming_language)) do not expose these pointers to the programmer.

Reference types have identity, meaning that it is possible to distinguish two references at runtime, even when they contain underlying values that are equal.

----

# *Essential C# 7.0*

## Chapter 3: More with Data Types

### Reference Types

By contrast. the value of a variable of reference type is a reference to a storage location that contains data. Reference types store the reference where the data is located instead of storing the data directly, as value types do.
Therefore, to access the data, the runtime reads the memory location out of the variable and the "jumps" to the location in memory that contains the data. The memory area of the data a reference type points to is called the heap.

----

# *[blog: rust-references-vs-pointers](https://ntietz.com/blog/rust-references-vs-pointers/)*

Ultimately, the underlying representation is the same: both hold an address for some memory. The difference between them is ultimately in semantics.

References have some [rules](https://doc.rust-lang.org/nomicon/references.html) enforced by the compiler. Specifically, they cannot outlive what they refer to (the "referent"), and mutable references cannot be aliased. Other than that, references behave a lot like the variables they point to. They have a type, and you can interact with that type to read it or (with mutable references) modify it.

On the other hand, pointers are semantically more about the address. This means that when we interact with them, we'll be modifying the address (things like add will do pointer offsets instead of adding to the underlying value). When we print them, we don't print the underlying value—in fact, we cannot get to the underlying value at all without the unsafe keyword. Instead, we print out the address.

<h2><strong>the difference between references and pointers in Rust: they have the same underlying data, but different constraints and semantics with the compiler.</strong></h2>

----

正如rust-references-vs-pointers这篇博客所言，指针和引用的底层实现是相同的，不同点在于规则约束和编译器的语义。
指针更偏向于操纵地址而非地址所指向的值，引用则反之。

在其他的语言描述中，指针类型被认为是不安全的。如rust文档中，是这样描述指针的：
```
Primitive Type pointer
Raw, unsafe pointers, *const T, and *mut T.
```

在golang中，指针类型直接被编入unsafe包中：
```
type Pointer *ArbitraryType

Pointer represents a pointer to an arbitrary type. There are four special operations available for type Pointer that are not available for other types:

A pointer value of any type can be converted to a Pointer.
A Pointer can be converted to a pointer value of any type.
A uintptr can be converted to a Pointer.
A Pointer can be converted to a uintptr.

Pointer therefore allows a program to defeat the type system and read and write arbitrary memory. It should be used with extreme care.
```

其他语言中，引用类型自带语法糖，会隐性解引用(implicit de-reference)。这一点同样起到保护对象数据的作用。

**Reply by zhangzhimin** [reply]
我们可以通过把指针控制在array, list等数据结构之内来避免大部分情况下直接使用指针. 

安全与否不止要看行为,  大部分时候行为是否清晰明确更为重要. 上面4条都是清晰明确的指针操作, 且明确存在应用场景.

**Comment by microctar** [comment]
在参考部分资料后，我并不认为，引用是历史包袱。相反，引用是出于保护数据(可以参考c中的union类型，数据操纵可以精确到bit)的考虑而被设计出来的。

**Comment by zhangzhimin** [comment]
这个问题可以持续关注, 如果遇到必须使用引用的场景再讨论

**Comment by microctar** [comment]
### https://nim-lang.org/docs/manual.html#types-reference-and-pointer-types

References (similar to pointers in other programming languages) are a way to introduce many-to-one relationships. This means different references can point to and modify the same location in memory (also called aliasing).

Nim distinguishes between traced and untraced references. Untraced references are also called pointers. Traced references point to objects of a garbage-collected heap, untraced references point to manually allocated objects or objects somewhere else in memory. Thus, untraced references are unsafe. However, for certain low-level operations (accessing the hardware) untraced references are unavoidable.

Traced references are declared with the ref keyword, untraced references are declared with the ptr keyword. In general, a ptr T is implicitly convertible to the pointer type.

**Reply by zhangzhimin** [reply]
Prajna有ptr也就是C语言的指针, 目前来看引用没有什么必须的场景. 
 FunCall(x)如果存在引用, 我们无法有效辨别x是否会被改变? 我们去除引用之后, x必然是一个值, 不会被改变. 但需要改变时我们传入x的指针.

**Reply by zhangzhimin** [reply]
但潜在的引用是存在的, 我们访问字段point.x=3其实x就是个等价于引用的东西, 这块可以看下Prajna里DeferencePointer, 通过对"解"指针来实现读写操作.
